File: backend/src/core/aegis/providers/gemini-provider.js
import { TaskType, ExecutionStatus } from '../task-types.js';
import { GeminiClient } from '../clients/gemini-client.js';

export class GeminiProvider {
  constructor({ client } = {}) {
    this.id = 'aegis-gemini';
    this.name = 'Google Gemini';
    this.taskType = TaskType.TEXT;
    this.isMock = false;
    this._client = client ?? new GeminiClient();
  }

  async execute(request, _classification) {
    const started = Date.now();
    const prompt = this._buildPrompt(request);

    try {
      const content = await this._client.generateWithRetry(prompt);
      return {
        taskType: this.taskType,
        providerId: this.id,
        providerName: this.name,
        executionStatus: ExecutionStatus.COMPLETED,
        content,
        isMock: false,
        durationMs: Date.now() - started,
      };
    } catch (error) {
      const message = this._formatError(error);
      return {
        taskType: this.taskType,
        providerId: this.id,
        providerName: this.name,
        executionStatus: ExecutionStatus.FAILED,
        content: message,
        errorMessage: message,
        isMock: false,
        durationMs: Date.now() - started,
      };
    }
  }

  _buildPrompt(request) {
    const base = String(request.prompt ?? '').trim();
    if (request.attachmentName && !base) {
      return `The user attached a file named "${request.attachmentName}". Explain how you would help.`;
    }
    if (request.attachmentName) {
      return `${base}\n\n[Attached file: ${request.attachmentName}]`;
    }
    return base;
  }

  _formatError(error) {
    if (error.code === 'INVALID_API_KEY') {
      return 'Invalid or missing GEMINI_API_KEY. Check your backend .env file.';
    }
    if (error.code === 'TIMEOUT') {
      return 'Gemini request timed out. Please try again.';
    }
    return `Gemini error: ${error.message}`;
  }
}

File: lib/core/aegis/providers/gemini_provider.dart
import '../clients/gemini_api_client.dart';
import '../clients/gemini_exceptions.dart';
import '../domain/ai_provider.dart';
import '../domain/classification_result.dart';
import '../domain/execution_status.dart';
import '../domain/provider_request.dart';
import '../domain/provider_response.dart';
import '../domain/task_type.dart';

class GeminiProvider implements AIProvider {
  GeminiProvider({GeminiApiClient? client}) : _client = client ?? GeminiApiClient();

  final GeminiApiClient _client;

  @override
  String get id => 'aegis-gemini';

  @override
  String get name => 'Google Gemini';

  @override
  TaskType get taskType => TaskType.text;

  @override
  bool get isMock => false;

  @override
  Future<ProviderResponse> execute(
    ProviderRequest request, {
    required ClassificationResult classification,
  }) async {
    final started = DateTime.now();
    final prompt = _buildPrompt(request);

    try {
      final content = await _client.generateWithRetry(prompt);
      return ProviderResponse(
        taskType: taskType,
        providerId: id,
        providerName: name,
        executionStatus: ExecutionStatus.completed,
        content: content,
        durationMs: DateTime.now().difference(started).inMilliseconds,
      );
    } on GeminiAuthException catch (e) {
      return _failed(e.message, started);
    } on GeminiTimeoutException catch (e) {
      return _failed(e.message, started);
    } on GeminiException catch (e) {
      return _failed(e.message, started);
    } catch (e) {
      return _failed('Unexpected Gemini error: $e', started);
    }
  }

  String _buildPrompt(ProviderRequest request) {
    final base = request.normalizedPrompt;
    if (request.attachmentName != null && base.isEmpty) {
      return 'The user attached a file named "${request.attachmentName}". '
          'Explain how you would help process it.';
    }
    if (request.attachmentName != null) {
      return '$base\n\n[Attached file: ${request.attachmentName}]';
    }
    return base;
  }

  ProviderResponse _failed(String message, DateTime started) {
    return ProviderResponse(
      taskType: taskType,
      providerId: id,
      providerName: name,
      executionStatus: ExecutionStatus.failed,
      content: message,
      errorMessage: message,
      durationMs: DateTime.now().difference(started).inMilliseconds,
    );
  }
}
